indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java

```diff
@@ -269,11 +302,90 @@ public void setConf(Configuration config)
     }
   }
 
-  public static class IndexGeneratorReducer extends Reducer<BytesWritable, Writable, BytesWritable, Text>
+  public static class IndexGeneratorCombiner extends Reducer<BytesWritable, BytesWritable, BytesWritable, BytesWritable>
+  {
+    protected HadoopDruidIndexerConfig config;
+    private AggregatorFactory[] combiningAggs;
+
+    @Override
+    protected void setup(Context context)
+        throws IOException, InterruptedException
+    {
+      config = HadoopDruidIndexerConfig.fromConfiguration(context.getConfiguration());
+
+      List<AggregatorFactory> combiningAggsList = Lists.newArrayList();
+      for (AggregatorFactory factory : config.getSchema().getDataSchema().getAggregators()) {
+        combiningAggsList.add(factory.getCombiningFactory());
+      }
+      combiningAggs = combiningAggsList.toArray(new AggregatorFactory[0]);
```

Comment:
This pattern can actually generate a *lot* of garbage.  The `add()` to the List will result in binary growth of the underlying array and then in the end ***another*** array is created to copy stuff into it.  While it won't hurt anything here, this is a fairly bad code pattern to follow in general.  Either way, in this case, I think the faster and less-wasteful code is even clearer:

```
AggregatorFactory[] aggs = config.getSchema().getDataSchema().getAggregators();
combiningAggs = new AggregatorFactory[aggs.length];
for (int i = 0; i < aggs.length; ++i) {
  combiningAggs[i] = aggs[i].getCombiningFactory();
}
```